smithy-model/src/main/java/software/amazon/smithy/model/node/IdentityClassCache.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.node;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Supplier;
+
+/**
+ * Caches instances of loaded classes of type T using referential equality
+ * to compare keys.
+ *
+ * <p>Cached values are invalidated if the requested class != currently cached
+ * class. This ensures that the most recently loaded versions of classes are
+ * being used.
+ *
+ * @param <T> The type of the cached classes.
+ */
+final class IdentityClassCache<T> {
+
+    private final ConcurrentMap<String, T> cache;
+
+    IdentityClassCache() {
```

Comment:
nit - you could just remove the ctor if there are no args and instantiate the cache on line 34

Reply:
Updated